File: app/src/main/java/org/dhis2/usescases/splash/SplashPresenter.java
package org.dhis2.usescases.splash;


import android.content.Context;
import android.content.SharedPreferences;
import android.content.res.Resources;

import org.dhis2.data.server.UserManager;
import org.dhis2.usescases.login.LoginActivity;
import org.dhis2.usescases.main.MainActivity;
import org.dhis2.usescases.sync.SyncActivity;
import org.dhis2.utils.Constants;
import org.dhis2.utils.SyncUtils;

import java.util.concurrent.TimeUnit;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.schedulers.Schedulers;
import timber.log.Timber;

import static android.text.TextUtils.isEmpty;

public class SplashPresenter implements SplashContracts.Presenter {

    private final SplashRepository splashRespository;
    private SplashContracts.View view;
    private UserManager userManager;
    @NonNull
    private final CompositeDisposable compositeDisposable;

    SplashPresenter(@Nullable UserManager userManager, SplashRepository splashRepository) {
        this.userManager = userManager;
        this.compositeDisposable = new CompositeDisposable();
        this.splashRespository = splashRepository;
    }

    @Override
    public void destroy() {
        compositeDisposable.clear();
    }

    @Override
    public void init(SplashContracts.View view) {
        this.view = view;

        compositeDisposable.add(splashRespository.getIconForFlag()
                .delay(2, TimeUnit.SECONDS, Schedulers.io())
                .map(flagName -> {
                    if (!isEmpty(flagName)) {
                        Resources resources = view.getAbstracContext().getResources();
                        return resources.getIdentifier(flagName, "drawable", view.getAbstracContext().getPackageName());
                    } else
                        return -1;
                })
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(
                        view.renderFlag(),
                        Timber::d
                )
        );
    }

    @Override
    public void isUserLoggedIn() {
        if (userManager == null) {
            navigateToLoginView();
            return;
        }

        if (SyncUtils.isSyncRunning()) {

            view.startActivity(SyncActivity.class, null, true, true, null);

        } else {

            compositeDisposable.add(userManager.isUserLoggedIn()
                    .delay(2000, TimeUnit.MILLISECONDS, Schedulers.io())
                    .subscribeOn(Schedulers.io())
                    .observeOn(AndroidSchedulers.mainThread())
                    .subscribe(isUserLoggedIn -> {
                        SharedPreferences prefs = view.getAbstracContext().getSharedPreferences(
                                Constants.SHARE_PREFS, Context.MODE_PRIVATE);
                        if (isUserLoggedIn && !prefs.getBoolean("SessionLocked", false)) {
                            navigateToHomeView();
                        } else {
                            navigateToLoginView();
                        }
                    }, Timber::e));
        }
    }

    @Override
    public void navigateToLoginView() {
        view.startActivity(LoginActivity.class, null, true, true, null);
    }

    @Override
    public void navigateToHomeView() {
        view.startActivity(MainActivity.class, null, true, true, null);
    }

}

File: app/src/main/java/org/dhis2/usescases/splash/SplashPresenter.kt
package org.dhis2.usescases.splash

import io.reactivex.disposables.CompositeDisposable
import org.dhis2.commons.prefs.Preference
import org.dhis2.commons.prefs.PreferenceProvider
import org.dhis2.commons.schedulers.SchedulerProvider
import org.dhis2.data.server.UserManager
import org.dhis2.mobile.commons.reporting.CrashReportController
import org.dhis2.utils.analytics.DATA_STORE_ANALYTICS_PERMISSION_KEY
import timber.log.Timber
import java.util.concurrent.TimeUnit

class SplashPresenter internal constructor(
    private var view: SplashView,
    private val userManager: UserManager?,
    private val schedulerProvider: SchedulerProvider,
    private val preferenceProvider: PreferenceProvider,
    private val crashReportController: CrashReportController,
) {
    var compositeDisposable: CompositeDisposable = CompositeDisposable()

    fun destroy() {
        compositeDisposable.clear()
    }

    fun init() {
        isUserLoggedIn()
    }

    private fun isUserLoggedIn() {
        userManager?.let { userManager ->
            compositeDisposable.add(
                userManager.isUserLoggedIn
                    .delay(2000, TimeUnit.MILLISECONDS, schedulerProvider.io())
                    .subscribeOn(schedulerProvider.io())
                    // Resolve the blocking SDK reads here, upstream of observeOn(ui): if they
                    // run after the switch to the UI scheduler they block the main thread on
                    // SQLiteConnectionPool while background sync holds the connection, causing
                    // an ANR (DHIS2-ANDROID-CAPTURE-83NF / 83P1).
                    .map { userLogged -> userLogged to trackingInfoFor(userManager, userLogged) }
                    .observeOn(schedulerProvider.ui())
                    .subscribe(
                        { (userLogged, trackingInfo) ->
                            trackingInfo?.let {
                                trackUserInfo(
                                    serverUrl = it.serverUrl,
                                    serverVersion = it.serverVersion,
                                )
                            }
                            view.goToNextScreen(
                                userLogged,
                                preferenceProvider.getBoolean(
                                    Preference.SESSION_LOCKED,
                                    false,
                                ),
                                preferenceProvider.getBoolean(
                                    Preference.INITIAL_METADATA_SYNC_DONE,
                                    false,
                                ),
                                preferenceProvider.getBoolean(
                                    Preference.INITIAL_DATA_SYNC_DONE,
                                    false,
                                ),
                            )
                        },
                        { Timber.d(it) },
                    ),
            )
        } ?: view.goToNextScreen(
            false,
            sessionLocked = false,
            initialSyncDone = false,
            initialDataSyncDone = false,
        )
    }

    private fun trackingInfoFor(
        userManager: UserManager,
        userLogged: Boolean,
    ): TrackingInfo? =
        if (userLogged && trackingPermissionGranted(userManager)) {
            val systemInfo =
                userManager.d2
                    .systemInfoModule()
                    .systemInfo()
                    .blockingGet()
            TrackingInfo(
                serverUrl = systemInfo?.contextPath() ?: "",
                serverVersion = systemInfo?.version() ?: "",
            )
        } else {
            null
        }

    private fun trackingPermissionGranted(userManager: UserManager): Boolean =
        userManager.d2
            .dataStoreModule()
            .localDataStore()
            .value(DATA_STORE_ANALYTICS_PERMISSION_KEY)
            .blockingGet()
            ?.value() == true.toString()

    private fun trackUserInfo(
        serverUrl: String,
        serverVersion: String,
    ) {
        crashReportController.trackServer(serverUrl, serverVersion)
    }

    fun getAccounts(): Int =
        userManager
            ?.d2
            ?.userModule()
            ?.accountManager()
            ?.getAccounts()
            ?.count() ?: 0

    private data class TrackingInfo(
        val serverUrl: String,
        val serverVersion: String,
    )
}
